src/darsia/utils/linear_solvers/mg.py

```diff
@@ -0,0 +1,235 @@
+"""
+Multigrid solver for linear systems arising from discretizations of the
+minimization problem. Could be used as a preconditioner or a solver in itself.
+"""
+
+
+from typing import Optional, Union
+
+import numpy as np
+
+import darsia as da
+
+
+class MG(da.Solver):
+    """
+    Muiltilvel solver for the problem: mass_coeff * x - diffusion_coeff * laplace(x) = rhs
+    """
+
+    def __init__(
+        self,
+        mass_coeff: Union[float, np.ndarray],
+        diffusion_coeff: Union[float, np.ndarray],
+        depth: int = 2,
+        smoother_iterations: int = 5,
+        dim: int = 2,
+        maxiter: int = 100,
+        tol: Optional[float] = None,
+        verbose=False,
+    ) -> None:
+        """
+        Initialize the solver.
+
+        Args:
+            mass_coeff [np.ndarray,float]: mass coefficient
+            diffusion_coeff [np.ndarray,float]: diffusion coefficient
+            depth [int]: depth of the multigrid hierarchy
+            smoother_iterations [int]: number of iterations for the smoother
+            dim [int]: dimension of the problem
+            maxiter [int]: maximum number of iterations
+            tol [Optional[float]]: tolerance
+            verbose [bool]: print information
+
+        """
+        super().__init__(
+            mass_coeff=mass_coeff,
+            diffusion_coeff=diffusion_coeff,
+            dim=dim,
+            maxiter=maxiter,
+            tol=tol,
+            verbose=verbose,
+        )
+        self.smoother = da.Jacobi(
+            mass_coeff=mass_coeff,
+            diffusion_coeff=diffusion_coeff,
+            dim=dim,
+            maxiter=smoother_iterations,
+        )
+        self.smoother_iterations = smoother_iterations
+        self.depth = depth
+        self.heterogeneous = False
+        if isinstance(mass_coeff, np.ndarray) or isinstance(
+            diffusion_coeff, np.ndarray
+        ):
+            self.heterogeneous = True
+
+    def operator(self, x: np.ndarray, h: float) -> np.ndarray:
+        """The solution operator for the problem
+
+        Args:
+            x [np.ndarray]: input
+            h [float]: grid spacing
+
+        Returns:
+            output [np.ndarray]
+
+        """
+
+        return self.mass_coeff * x - self.diffusion_coeff * da.laplace(
+            x, dim=self.dim, h=h
+        )
+
+    def restriction(self, x: np.ndarray) -> np.ndarray:
+        """Restrict x, i.e., coarsen it by a factor 2. Even and odd indices are averaged.
```

Comment:
What if `x` does not have an even size?

Reply:
Then we lose the last row/column/..

Reply:
OK. Should be OK. Maybe add a comment explicitly stating this?